### ISSUE TITLE
Issue with default_value for List trait

### ISSUE THREAD

**Issue by ssunkara1** [ipython/traitlets#366]
I can't seem to set `default_value` of `List` trait to be `None`. When I try to set it the using the `default_value` argument in the constructor, the value is set to `[]`. After initialization I can set it to `None`. I am using `traitlets 4.3.1`. It works when the trait is an `Instance` of `List`.

```python
from traitlets import List, Instance
from ipywidgets import DOMWidget

class Test(DOMWidget):
    arr = Instance(List, default_value=None, allow_none=True).tag(sync=True)
    arr_2 = List(default_value=None, allow_none=True).tag(sync=True)
    
test_obj = Test()
print(test_obj.arr is None)
print(test_obj.arr_2 is None)
print(test_obj.arr_2)
```

Output:
True
False
[]

**Comment by SylvainCorlay** [comment]
Which version of traitlets?

**Comment by ssunkara1** [comment]
traitlets 4.3.1

**Comment by tonyfast** [comment]
I don't know if this is related, but I am having issues with the `default_value` on `Instance`.  This is using the current `master` .

```python
class Foo(traitlets.HasTraits):
    bar = traitlets.Instance('jinja2.Environment', default_value=jinja2.Environment())
    
print(type(Foo.bar.default()))
print(hasattr(Foo.bar, 'make_dynamic_default'))
print(getattr(Foo.bar, 'make_dynamic_default')())
```
gives
```
<class 'NoneType'>
True
None
````

It might be that [`make_dynamic_defaults`](https://github.com/ipython/traitlets/blob/master/traitlets/traitlets.py#L489) should be checked for `None`.

**Comment by rmorshea** [comment]
I am reasonably certain that the original error results [here](https://github.com/ipython/traitlets/blob/master/traitlets/traitlets.py#L2244) on master, but was also present [here](https://github.com/ipython/traitlets/blob/4.0/traitlets/traitlets.py#L1520) in the 4.0 release. I don't know if we should considered this a bug...

@tonyfast #367 fixes the bug you found which appears to be unrelated in function the original.

**Comment by SylvainCorlay** [comment]
Yes, the original container class has a convoluted logic. I wanted to change this when we did the traitlets refactoring but we had to stop at some point.

**Comment by SylvainCorlay** [comment]
Yes, the original container class has a convoluted logic. I wanted to change this when we did the traitlets refactoring but we had to stop at some point.

**Comment by rmorshea** [comment]
@SylvainCorlay are you saying this is a change you want to make, or a change that broke too many things for you to go forward with it?
